### PR TITLE
importinto: make cancel wait task done and some fixes

### LINKED ISSUE #48736: cancel
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### PR DESCRIPTION
**Pull request by D3Hunter**
### What problem does this PR solve?

Issue Number: close #48736

Problem Summary:

### What changed and how does it work?

- remove OnErrStage and add OnDone to dispatcher
- let dist framework manage state of import_jobs
- remove `truncate table` on error
- `cancel import job` now wait task done
- fix bug that `row2GlobeTask` uses marshal error as task error
- fix bug that `GetTaskImportedRows` didn't consider task done in middle
- some refine

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/disttask/framework/dispatcher/dispatcher.go [hunk 1]
```diff
@@ -544,7 +547,14 @@ func (d *BaseDispatcher) onNextStage() (err error) {
...
 				taskState = proto.TaskStateSucceed
-				logutil.Logger(d.logCtx).Info("all subtasks dispatched and processed, finish the task")
+				logutil.Logger(d.logCtx).Info("task succeed",
+					zap.Stringer("type", d.Task.Type),
+					zap.Int64("task-id", d.Task.ID),
+				)
+				if err = d.OnDone(d.ctx, d, d.Task); err != nil {
+					err = errors.Trace(err)
+					return
+				}
```
Comment (ywqzzy) [inline pkg/disttask/framework/dispatcher/dispatcher.go:557]: If onDone failed, we will see "task succeed" multiple times, is it expected?

Comment (D3Hunter) [reply pkg/disttask/framework/dispatcher/dispatcher.go:557]: this log seems not that useful, will delete it

File: pkg/disttask/framework/dispatcher/interface.go [hunk 2]
```diff
@@ -66,10 +66,11 @@ type Extension interface {
...
 
-	// OnErrStage is called when:
-	// 	1. subtask is finished with error.
-	// 	2. task is cancelled after we have dispatched some subtasks.
-	OnErrStage(ctx context.Context, h TaskHandle, task *proto.Task, receiveErrs []error) (subtaskMeta []byte, err error)
+	// OnDone is called when task is done, either finished successfully or failed
+	// with error.
+	// if the task is failed when initializing dispatcher, or it's an unknown task,
+	// we don't call this function.
+	OnDone(ctx context.Context, h TaskHandle, task *proto.Task) error
```
[the remark this answers is not in the document]
Comment (D3Hunter) [reply pkg/disttask/framework/dispatcher/interface.go:73]: most methods in this interface should be, we just didn't state it explicitly

File: pkg/disttask/framework/framework_err_handling_test.go [hunk 3]
```diff
@@ -101,8 +101,8 @@ func (p *planNotRetryableErrDispatcherExt) OnNextSubtasksBatch(_ context.Context
 	return nil, errors.New("not retryable err")
 }
 
-func (*planNotRetryableErrDispatcherExt) OnErrStage(_ context.Context, _ dispatcher.TaskHandle, _ *proto.Task, _ []error) (meta []byte, err error) {
-	return nil, errors.New("not retryable err")
```
Comment (okJiang) [inline pkg/disttask/framework/framework_err_handling_test.go:105]: delete accidentally?

Comment (D3Hunter) [reply pkg/disttask/framework/framework_err_handling_test.go:105]: OnErrStage is removed

File: pkg/disttask/framework/dispatcher/dispatcher.go (hunk 1 above)
Comment (D3Hunter) [reply pkg/disttask/framework/dispatcher/dispatcher.go]: https://github.com/pingcap/tidb/pull/48928/commits/66f294d9bd670c923775be91654f16c39d32f4f5

File: pkg/disttask/framework/dispatcher/dispatcher.go [hunk 4]
```diff
@@ -492,7 +488,7 @@ func (d *BaseDispatcher) dispatchSubTask4Revert(meta []byte) error {
 		subTasks = make([]*proto.Subtask, 0, len(instanceIDs))
 		for _, id := range instanceIDs {
 			// reverting subtasks belong to the same step as current active step.
-			subTasks = append(subTasks, proto.NewSubtask(d.Task.Step, d.Task.ID, d.Task.Type, id, meta))
+			subTasks = append(subTasks, proto.NewSubtask(d.Task.Step, d.Task.ID, d.Task.Type, id, []byte("{}")))
```
Comment (ywqzzy) [inline pkg/disttask/framework/dispatcher/dispatcher.go:491]: Will business need subtask meta to rollback?
ok for now.

Comment (D3Hunter) [reply pkg/disttask/framework/dispatcher/dispatcher.go:491]: might be, but we don't have real rollback now, and current rollback design that generate a task on each node looks bad to me.

maybe we can delete it completely later

**Review by ywqzzy — APPROVED** [review]

File: pkg/disttask/framework/dispatcher/dispatcher.go (hunk 4 above)
Comment (ywqzzy) [reply pkg/disttask/framework/dispatcher/dispatcher.go:491]: LGTM

Comment (ywqzzy) [reply pkg/disttask/framework/dispatcher/dispatcher.go:491]: > might be, but we don't have real rollback now, and current rollback design that generate a task on each node looks bad to me, rollback shouldn't depends on node
> 
> maybe we can delete it completely later

Maybe add some cleanup logic for each nodes' scheduler is better.

**Review by okJiang — APPROVED** [review]

File: pkg/disttask/framework/dispatcher/dispatcher.go (hunk 4 above)
Comment (D3Hunter) [reply pkg/disttask/framework/dispatcher/dispatcher.go:491]: what to clean? disk occupied by subtask? this should be done before subtask success/fail

**Pull request merged**